rasa/utils/tensorflow/rasa_layers.py

```diff
@@ -0,0 +1,423 @@
+
+            if self.use_sparse_dropout:
+                self._sparse_dropout = layers.SparseDropout(rate=dropout_rate)
+
+        if self.use_dense_dropout:
+            self._dense_dropout = tf.keras.layers.Dropout(rate=dropout_rate)
+
+    def call(
+        self,
+        features: List[Union[tf.Tensor, tf.SparseTensor]],
+        training: Optional[Union[tf.Tensor, bool]] = None,
+    ) -> Optional[tf.Tensor]:
+        dense_features = []
+        for f in features:
+            if isinstance(f, tf.SparseTensor):
+                if self.use_sparse_dropout:
+                    _f = self._sparse_dropout(f, training)
+                else:
+                    _f = f
+
+                dense_f = self._sparse_to_dense(_f)
+
+                if self.use_dense_dropout:
+                    dense_f = self._dense_dropout(dense_f, training)
+
+                dense_features.append(dense_f)
+            else:
+                dense_features.append(f)
+
+        return tf.concat(dense_features, axis=-1)
+
+
+class ConcatenateSequenceSentenceFeatures(tf.keras.layers.Layer):
+    # TODO add docstring
+    def __init__(
+        self,
+        layer_name_suffix: Text,
+        concat_dimension: int,
+        sequence_signature: FeatureSignature,
+        sentence_signature: FeatureSignature,
+        concat_layers_kwargs: Dict[Text, Any] = {},
+    ) -> None:
+        super().__init__(
+            name=f"concatenate_sequence_sentence_features_{layer_name_suffix}"
+        )
+        if sequence_signature and sentence_signature:
+            self.do_concatenation = True
+            if sequence_signature.units != sentence_signature.units:
+                self.unify_dimensions_before_concat = True
+                self.output_units = concat_dimension
+                self.unify_dimensions_layers = {}
+                for feature_type in [SEQUENCE, SENTENCE]:
+                    if "layer_name_suffix" not in concat_layers_kwargs:
+                        concat_layers_kwargs[
+                            "layer_name_suffix"
+                        ] = f"unify_dimensions_before_concat.{layer_name_suffix}_{feature_type}"
+                    self.unify_dimensions_layers[feature_type] = layers.Ffnn(
+                        **concat_layers_kwargs
+                    )
+            else:
+                self.unify_dimensions_before_concat = False
+                self.output_units = sequence_signature.units
+        else:
+            self.do_concatenation = False
+            if sequence_signature and not sentence_signature:
+                self.return_just = SEQUENCE
+                self.output_units = sequence_signature.units
+            elif sentence_signature and not sequence_signature:
+                self.return_just = SENTENCE
+                self.output_units = sentence_signature.units
+
+    def call(
+        self, sequence_x: tf.Tensor, sentence_x: tf.Tensor, mask_text: tf.Tensor,
```

Comment:
I don't think `_x` is needed

Reply:
Sure, I agree, I just copy-pasted this from `models.py`.